### PR TITLE
[ZkTracer] Nightly test sampling

### PR DESCRIPTION
**Pull request by OlivierBBB**
Implements https://github.com/Consensys/linea-monorepo/issues/2578

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes nightly test coverage by introducing deterministic-per-day sampling; failures may become less reproducible across days and some edge cases may be skipped compared to full enumeration.
> 
> **Overview**
> Reduces the runtime of several *nightly* parameterized tests by generating the full Cartesian-product argument lists and then running a day-of-month–seeded random subset (e.g., `ExpTest`, `MxpxThresholdTests`, `LowGasStipendPrecompileCallTests`, `XbsLimitsTests`).
> 
> Adds sampling-size constants per test suite and refactors `@MethodSource` providers to return `List<Arguments>` for full sets plus sampled `Stream<Arguments>` wrappers. Also hardens `Utilities.randomSampleByDayOfMonth` to cap the sample size to the available items to avoid `subList` bounds errors.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c9e4e0701f9918d3f754aea49852a60021fa81e1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java [hunk 1]
```diff
@@ -239,13 +255,12 @@ static Stream<Arguments> lowGasStipendPrecompileCallP256TestSource() {
         arguments.add(Arguments.of(P256_VERIFY, valueCase, gasCase, Integer.MAX_VALUE, false));
       }
     }
-    arguments.clear();
     arguments.add(Arguments.of(P256_VERIFY, ValueCase.ZERO, COST_MINUS_ONE, 160, false));
     arguments.add(Arguments.of(BLS12_G1ADD, ValueCase.ZERO, COST, 13, false));
-    return arguments.stream();
+    return arguments;
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java:258]: @lorenzogentile404 was the `arguments.clear()` there on purpose ?

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java [hunk 2]
```diff
@@ -239,13 +255,13 @@ static Stream<Arguments> lowGasStipendPrecompileCallP256TestSource() {
         arguments.add(Arguments.of(P256_VERIFY, valueCase, gasCase, Integer.MAX_VALUE, false));
       }
     }
-    arguments.clear();
-    arguments.add(Arguments.of(P256_VERIFY, ValueCase.ZERO, COST_MINUS_ONE, 160, false));
-    arguments.add(Arguments.of(BLS12_G1ADD, ValueCase.ZERO, COST, 13, false));
-    return arguments.stream();
+    // arguments.clear();
```
Comment (amkCha) [inline tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java]: Should we just remove that ?

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java [hunk 3]
```diff
@@ -213,7 +215,21 @@ void lowGasStipendPrecompileCallTest(
...
+        LOW_GAS_STIPEND_PRECOMPILE_TEST_SAMPLE_SIZE, fullGasStipendPrecompileCallP256TestSource())
+        .stream();
+  }
+
+  static List<Arguments> fullGasStipendPrecompileCallP256TestSource() {
+    List<Arguments> arguments = new ArrayList<>();
+    arguments.addAll(lowGasStipendPrecompileCallTestSource());
+    arguments.addAll(lowGasStipendPrecompileCallP256TestSource());
+
+    return arguments;
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/LowGasStipendPrecompileCallTests.java:229]: Valid but won't change it.

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpxThresholdTests.java [hunk 4]
```diff
@@ -284,18 +301,20 @@ private static Stream<Arguments> inputs(List<Bytes32> inputsValues) {
...
+    return randomSampleByDayOfMonth(MXPX_THRESHOLD_SAMPLE_SIZE, inputs(inputsValuesNightly))
+        .stream();
   }
 
-  private static Stream<Arguments> inputParamsUnit() {
-    return inputs(inputsValuesUnit);
+  private static Stream<Arguments> sampleInputParamsUnit() {
+    return randomSampleByDayOfMonth(MXPX_THRESHOLD_SAMPLE_SIZE, inputs(inputValuesUnitTests))
+        .stream();
   }
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpxThresholdTests.java]: Done

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/osakaModexpTests/XbsLimitsTests.java [hunk 5]
```diff
@@ -139,7 +142,12 @@ private void body(
                                                       bbsType, ebsType, mbsType))))))
           .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
 
-  static Stream<Arguments> modexpXbsLimitsTestsNighlySource() {
+  static Stream<Arguments> sampleModexpXbsLimitsTestsNighlySource() {
+    return randomSampleByDayOfMonth(XBS_LIMIT_TEST_SAMPLE_SIZE, modexpXbsLimitsTestsNighlySource())
+        .stream();
+  }
+
+  static List<Arguments> modexpXbsLimitsTestsNighlySource() {
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/precompiles/osakaModexpTests/XbsLimitsTests.java]: Done

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/callTests/Utilities.java [hunk 6]
```diff
@@ -343,6 +343,6 @@ public static <T> List<T> randomSampleByDayOfMonth(int n, List<T> items) {
     // Randomly shuffle the items
     Collections.shuffle(items, rng);
     //
-    return items.subList(0, n);
+    return items.subList(0, Math.min(n, items.size()));
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/callTests/Utilities.java:346]: Not an issue at the moment.

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpxThresholdTests.java [hunk 7]
```diff
@@ -284,18 +301,20 @@ private static Stream<Arguments> inputs(List<Bytes32> inputsValues) {
...
+  static Stream<Arguments> sampleInputParamsNightly() {
+    return randomSampleByDayOfMonth(MXPX_THRESHOLD_SAMPLE_SIZE, inputs(inputsValuesNightly))
+        .stream();
   }
 
-  private static Stream<Arguments> inputParamsUnit() {
-    return inputs(inputsValuesUnit);
+  private static Stream<Arguments> sampleInputParamsUnit() {
+    return randomSampleByDayOfMonth(MXPX_THRESHOLD_SAMPLE_SIZE, inputs(inputValuesUnitTests))
+        .stream();
```
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpxThresholdTests.java]: Done

File: tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/callTests/Utilities.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (OlivierBBB) [reply tracer/arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/callTests/Utilities.java:346]: Not an issue atm.
